org.eclipse.languageserver/src/org/eclipse/languageserver/LanguageServiceAccessor.java

```diff
@@ -107,18 +111,57 @@ public boolean equals(Object obj) {
 	private static Map<WrapperEntryKey, ProjectSpecificLanguageServerWrapper> projectServers = new HashMap<>();
 
 	public static class LSPDocumentInfo {
+		@NonNull
+		private final URI fileUri;
+		private final IFile file;
+		private final IDocument document;
+		@NonNull
+		private final LanguageClientEndpoint languageClient;
 
-		public final URI fileUri;
-		public final IFile file;
-		public final IDocument document;
-		public final LanguageClientEndpoint languageClient;
+		public static LSPDocumentInfo create(URI fileUri, IFile file, IDocument document, LanguageClientEndpoint languageClient) {
+			if (fileUri == null || languageClient == null) {
+				return null;
+			}
+			return new LSPDocumentInfo(fileUri, file, document, languageClient);
+		}
 
-		public LSPDocumentInfo(URI fileUri, IFile file, IDocument document, LanguageClientEndpoint languageClient) {
+		private LSPDocumentInfo(@NonNull URI fileUri, IFile file, IDocument document, @NonNull LanguageClientEndpoint languageClient) {
```

Comment:
I see. Then I should review the logic in the two LanguageServiceAccessor.getLSPDocumentInfoFor methods, who are at this moment the only places where a LSPDocumentInfo is constructed. 

Maybe you can help with some suggestions. The code is what was already written, and doesn't cover all paths: there still are some TODOs in there, making it difficult to please the compiler. What do you think getLSPDocumentInfoFor should do when fileUri or languageClient are null? Return null or throw exception?